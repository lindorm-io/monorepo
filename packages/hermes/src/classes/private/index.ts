export * from "./HermesMetadata";
export * from "./HermesRegistry";
export * from "./HermesScanner";
