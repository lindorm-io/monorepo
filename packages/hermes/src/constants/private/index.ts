export * from "./checksum-store";
export * from "./domain";
export * from "./event-store";
export * from "./saga-store";
export * from "./view-store";
