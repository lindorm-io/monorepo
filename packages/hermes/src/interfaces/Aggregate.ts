import { ClassLike, Dict } from "@lindorm/types";
import { AggregateData } from "../types";
import { IHermesMessage } from "./HermesMessage";

export interface IAggregate<S extends Dict = Dict> extends AggregateData {
  apply(causation: IHermesMessage, event: ClassLike): Promise<void>;
  load(event: IHermesMessage): void;
  toJSON(): AggregateData<S>;
}
