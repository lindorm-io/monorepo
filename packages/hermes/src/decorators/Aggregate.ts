import { Constructor } from "@lindorm/types";
import { AggregateDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Aggregate(options: AggregateDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addAggregate({
      encryption: options.encryption || false,
      name: options.name || name,
      namespace: options.namespace || null,
      target: target as unknown as Constructor,
    });
  };
}
