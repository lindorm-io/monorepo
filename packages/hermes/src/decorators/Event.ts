import { Constructor } from "@lindorm/types";
import { EventDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Event(options?: EventDecoratorOptions): ClassDecorator {
  return function (target) {
    const { name, version } = extractNameData(target.name);
    globalHermesMetadata.addEvent({
      name: options?.name || name,
      target: target as unknown as Constructor,
      version: options?.version || version,
    });
  };
}
