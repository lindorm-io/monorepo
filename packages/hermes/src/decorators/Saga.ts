import { isArray } from "@lindorm/is";
import { Constructor } from "@lindorm/types";
import { SagaDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Saga(
  AggregateClass: Constructor | Array<Constructor>,
  options: SagaDecoratorOptions = {},
): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addSaga({
      aggregates: isArray(AggregateClass) ? AggregateClass : [AggregateClass],
      name: options.name || name,
      namespace: options.namespace || null,
      target: target as unknown as Constructor,
    });
  };
}
