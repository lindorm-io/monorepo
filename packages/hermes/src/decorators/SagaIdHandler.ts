import { Constructor } from "@lindorm/types";
import { SagaIdHandlerDescriptor } from "../types";
import { globalHermesMetadata } from "../utils/private";

export function SagaIdHandler<C extends Constructor>(
  EventClass: C,
): SagaIdHandlerDescriptor<C> {
  return function (target, key, descriptor) {
    globalHermesMetadata.addHandler({
      conditions: null,
      decorator: "SagaIdHandler",
      encryption: false,
      handler: descriptor.value,
      key: key.toString(),
      schema: null,
      target: target.constructor as Constructor,
      trigger: EventClass,
    });
  };
}
