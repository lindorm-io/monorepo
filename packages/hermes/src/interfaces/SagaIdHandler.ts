import { Constructor } from "@lindorm/types";
import { HandlerIdentifier, SagaIdCallback } from "../types";
import { NameData } from "../utils/private";

export interface ISagaIdHandler<C extends Constructor = Constructor> {
  aggregate: HandlerIdentifier;
  event: NameData;
  saga: HandlerIdentifier;
  handler: SagaIdCallback<C>;
}
