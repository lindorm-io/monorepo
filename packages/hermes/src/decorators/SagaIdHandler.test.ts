import { SagaIdCtx } from "../types";
import { globalHermesMetadata } from "../utils/private";
import { Aggregate } from "./Aggregate";
import { Event } from "./Event";
import { Saga } from "./Saga";
import { SagaIdHandler } from "./SagaIdHandler";

describe("SagaIdHandler Decorator", () => {
  @Aggregate()
  class TestSagaAggregate {}

  test("should add metadata", () => {
    @Event()
    class TestSagaIdentityHandlerEvent {
      public constructor(public readonly one: string) {}
    }

    @Saga(TestSagaAggregate)
    class TestSagaIdentityHandlerSaga {
      @SagaIdHandler(TestSagaIdentityHandlerEvent)
      public getIdTestSagaIdentityHandlerEvent(
        ctx: SagaIdCtx<TestSagaIdentityHandlerEvent>,
      ) {
        return ctx.event.one;
      }
    }

    expect(globalHermesMetadata.getSaga(TestSagaIdentityHandlerSaga)).toMatchSnapshot();
  });
});
