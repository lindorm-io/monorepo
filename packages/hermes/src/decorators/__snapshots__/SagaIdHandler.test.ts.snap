// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SagaIdHandler Decorator should add metadata 1`] = `
{
  "aggregates": [
    [Function],
  ],
  "handlers": [
    {
      "conditions": null,
      "decorator": "SagaIdHandler",
      "encryption": false,
      "handler": [Function],
      "key": "getIdTestSagaIdentityHandlerEvent",
      "schema": null,
      "target": [Function],
      "trigger": [Function],
    },
  ],
  "name": "test_saga_identity_handler_saga",
  "namespace": null,
  "target": [Function],
}
`;
